codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/ProtocolGenerator.java

```diff
@@ -252,90 +258,171 @@ static String getDeserializeMiddlewareName(ShapeId operationShapeId, ServiceShap
+     * @param context   the generation context
      * @param operation the operation shape to retrieve errors for
      * @return map of error names to {@link ShapeId}
      */
     default Map<String, ShapeId> getOperationErrors(GenerationContext context, OperationShape operation) {
         return HttpProtocolGeneratorUtils.getOperationErrors(context, operation);
     }
 
+    default void generateProtocolDocumentMarshalerUnmarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateProtocolDocumentMarshalerMarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateProtocolDocumentUnmarshalerUnmarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateProtocolDocumentUnmarshalerMarshalDocument(GenerationContext context) {
+        throw new CodegenException("document types not implemented for protocol");
+    }
+
+    default void generateNewDocumentMarshaler(GenerationContext context, Symbol marshalerSymbol) {
+        GoWriter writer = context.getWriter().get();
+        writer.openBlock("return &$T{", "}", marshalerSymbol, () -> {
+            writer.write("value: v,");
+        });
+    }
+
+    default void generateNewDocumentUnmarshaler(GenerationContext context, Symbol unmarshalerSymbol) {
+        GoWriter writer = context.getWriter().get();
+        writer.openBlock("return &$T{", "}", unmarshalerSymbol, () -> {
+            writer.write("value: v,");
+        });
+    }
+
     /**
      * Context object used for service serialization and deserialization.
      */
     class GenerationContext {
-        private GoSettings settings;
-        private Model model;
-        private ServiceShape service;
-        private SymbolProvider symbolProvider;
-        private GoWriter writer;
-        private List<GoIntegration> integrations;
-        private String protocolName;
-        private GoDelegator delegator;
+        private final GoSettings settings;
+        private final Model model;
+        private final ServiceShape service;
+        private final SymbolProvider symbolProvider;
+        private final GoWriter writer;
+        private final List<GoIntegration> integrations;
+        private final String protocolName;
+        private final GoDelegator delegator;
+
+        private GenerationContext(Builder builder) {
+            this.settings = SmithyBuilder.requiredState("settings", builder.settings);
+            this.model = SmithyBuilder.requiredState("model", builder.model);
+            this.service = SmithyBuilder.requiredState("service", builder.service);
+            this.symbolProvider = SmithyBuilder.requiredState("symbolProvider", builder.symbolProvider);
+            this.writer = builder.writer;
```

Comment:
Why is the writer optional?

Reply:
I refer to this in an earlier comment, but since the generation context is a bit of a grab bag, the writer is not required in all instances of the context's usage, thus is why it is optional.